### ISSUE TITLE
Problem with handling Chinese punctuation, such as enumeration comma(、) and full stop(。).

### ISSUE THREAD

**Issue by gapplef** [canonical/Ubuntu-Sans-fonts#117]
Some Chinese characters, such as enumeration comma(、) and full stop(。), displayed incorrectly in text-editor including gnome-text-editor, zim-wiki, and sublime-text etc. on Ubuntu 24.04, while in firefox everything displayed without problem.

Most special case is vscode, it have the same problem when use Chinese punctuation with English words, while used with Chinese words it could auto correct the display problem. The following picture shown the display of `测试、test、1、2、3、4、5、6。` and `test、测试、1、2、3、4、5、6。` in gnome-text-editor(left) and vscode(right).

![bug](https://github.com/canonical/Ubuntu-Sans-fonts/assets/5604665/55f0afda-6805-48b9-853b-46fce6ee27af)

The problem can be solved by setting custom font to `Monospace` or `Noto Sans CJK` in perferences of gnome-text-editor, while with `Ubuntu`, `Ubuntu Mono` or `Ubuntu Sans`, the problem exist. So this should be a problem related to the default font for Chinese characters in Ubuntu 24.04.

**Comment by rschiang** [comment]
Tested Ubuntu Sans 1.006 on macOS; seems fine. I would suggest looking into your font fallback stack or gnome’s rendering engine instead.
<img width="449"  src="https://github.com/canonical/Ubuntu-Sans-fonts/assets/1512231/bc56e48a-ff62-4a4b-a12a-75d84901df1f">

**Comment by gapplef** [comment]
@rschiang Thanks for your help!

Sorry, I am just a regular user meet this problem after install Ubuntu 24.04 recently. What do you mean by `font fallback stack or gnome’s rendering engine`? 

Are you suggesting there is something wrong with the font setting of Ubuntu 24.04 or GTK3&GTK4 font rendering?  It happened after a fresh installation, and I don't think I have touched any of those things, so it should be a bug somewhere. How could I check those and where should I report the bug?

- - - 
Also, I think the full stop(。) position is **not** correct in your screenshot. It should located at the bottom of the line, not the center, as you can see from the bottom right of my previous picture.

**Comment by rschiang** [comment]
Hi @gapplef, what I meant is that the problem doesn’t seem to be related to the Ubuntu font itself. Reading from your screenshot, it seems that either there are some problems regarding how your OS render these fonts, or there might be some bugs related to Gnome text editor. 

As you can see, the font correctly renders on other operating systems (in my case, MacOS). Hence, for others to assist you with the issue, we’ll probably need more information regarding how you set up the system or how we could reproduce the issue using a fresh Ubuntu copy.

---

Regarding to the question on full stop (。) position, no—the screenshot was perfectly fine. Folks in different parts of Asia uses different punctuation rules, as you could read from [Wikipedia](https://en.wikipedia.org/wiki/Full_stop#Chinese_and_Japanese):

> When used with [traditional characters](https://en.wikipedia.org/wiki/Traditional_Chinese_characters), the full stop is generally centered on the [mean line](https://en.wikipedia.org/wiki/Mean_line); when used with [simplified characters](https://en.wikipedia.org/wiki/Simplified_Chinese_characters), it is usually aligned to the baseline.

Hope this helps!

**Comment by gapplef** [comment]
As same problem also found in zim-wiki, sublime-text and vscode, so it not a  bug related to Gnome text editor. And the problem seems also not relate to gtk or gnome. https://gitlab.gnome.org/GNOME/gnome-text-editor/-/issues/675


It should be something relate to default font setting of my Ubuntu 24.04, I may need to check with a fresh installation when having time. The only special thing might relate to font problem is that I install ibus-rime and used [rime-ice](https://github.com/iDvel/rime-ice) config. Also I migrate some config dotfiles direct from Ubuntu 22.04, not sure is that relevant.

**Comment by gapplef** [comment]
@rschiang I can reproduce this bug with a fresh installation.
The problem only appear when the system language is set to English.
All thing is fine if set the language to Chinese.

**Comment by rschiang** [comment]
@gapplef I installed the 1.006 variable font on GNOME Editor on Fedora 40 and still cannot reproduce the issue. I haven’t had time to verify this on a fresh Ubuntu installation though. (Might do that during weekends)

Could this be an issue related to `zh-CN` fallback font? Could you examine your version of Noto Sans SC / Source Han Sans and see if it’s too old?
<img width="607" alt="Gnome Editor on Fedora 40" src="https://github.com/canonical/Ubuntu-Sans-fonts/assets/1512231/2b76cc7f-a712-43e0-a955-9e64f7e26f20">

**Comment by gapplef** [comment]
I'm not sure how to check the version, but here is the result of `apt search`:
![Screenshot from 2024-07-02 20-55-24](https://github.com/canonical/Ubuntu-Sans-fonts/assets/5604665/8184265b-4bcb-4492-8d63-9c7c3a2f576c)

**Comment by homanw104** [comment]
The CJK punctuations used _Noto Sans Mongolian_ as the fallback font. I fixed it by adding the following configuration to `/etc/fonts/local.conf`.

```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>

  <alias>
    <family>sans-serif</family>
    <prefer>
      <family>Noto Sans</family>
      <family>Noto Sans CJK SC</family>
      <family>Noto Sans CJK TC</family>
      <family>Noto Sans CJK HK</family>
      <family>Noto Sans CJK JP</family>
      <family>Noto Sans CJK KR</family>
    </prefer>
  </alias>
  <alias>
    <family>serif</family>
    <prefer>
      <family>Noto Serif</family>
      <family>Noto Serif CJK SC</family>
      <family>Noto Serif CJK TC</family>
      <family>Noto Serif CJK HK</family>
      <family>Noto Serif CJK JP</family>
      <family>Noto Serif CJK KR</family>
    </prefer>
  </alias>

</fontconfig>
```

The above configuration is referenced by `/etc/fonts/conf.d/51-local.conf` which has a higher priority than `56-language-selector-prefer.conf` where _Noto Sans Mongolian_ is preferred over other CJK fonts.

**Comment by devymex** [comment]
@homanw104 Thank you for your advice. It really helped solve my problem.

**Comment by gapplef** [comment]
@homanw104  put the same configuration under `~/.config/fontconfig/conf.d` also fixed my problem. Thanks a lot!

**Comment by wipstones** [comment]
@homanw104 I found that directly adding `/etc/fonts/local.conf` may cause some system UI fonts to change, such as the clock in top bar. Based on my tests, the fonts causing changes in Chinese punctuation are `Noto Sans Mongolian`, `Noto Sans PhagsPa`, `Noto Sans Tai Le` and `Noto Sans Yi`.  Commenting them out in `/etc/fonts/conf.d/56-language-selector-prefer.conf` will solve the issue.

```
...
84       <family>Noto Sans Miao</family>
85       <family>Noto Sans Modi</family>
86       <!--family>Noto Sans Mongolian</family-->
87       <family>Noto Sans Mro</family>
88       <family>Noto Sans Multani</family>
...
```
```
...
101       <family>Noto Sans Palmyrene</family>
102       <family>Noto Sans Pau Cin Hau</family>
103       <!--family>Noto Sans PhagsPa</family-->
104       <family>Noto Sans Phoenician</family>
105       <family>Noto Sans Rejang</family>
...
```
```
...
118       <family>Noto Sans Tagalog</family>
119       <family>Noto Sans Tagbanwa</family>
120       <!--family>Noto Sans Tai Le</family-->
121       <family>Noto Sans Tai Tham</family>
122       <family>Noto Sans Tai Viet</family>
...
```
```
...
133       <family>Noto Sans Wancho</family>
134       <family>Noto Sans Warang Citi</family>
135       <!--family>Noto Sans Yi</family-->
136       <family>Noto Sans Zanabazar Square</family>
137 
...
```

**Comment by devymex** [comment]
> @homanw104 I found that directly adding `/etc/fonts/local.conf` may cause some system UI fonts to change, such as the clock in top bar. Based on my tests, the fonts causing changes in Chinese punctuation are `Noto Sans Mongolian`, `Noto Sans PhagsPa`, and `Noto Sans Tai Le`. Commenting them out in `/etc/fonts/conf.d/56-language-selector-prefer.conf` will solve the issue.
> 
> ```
> ...
> 84       <family>Noto Sans Miao</family>
> 85       <family>Noto Sans Modi</family>
> 86       
> 87       <family>Noto Sans Mro</family>
> 88       <family>Noto Sans Multani</family>
> ...
> ```
> 
> ```
> ...
> 101       <family>Noto Sans Palmyrene</family>
> 102       <family>Noto Sans Pau Cin Hau</family>
> 103       
> 104       <family>Noto Sans Phoenician</family>
> 105       <family>Noto Sans Rejang</family>
> ...
> ```
> 
> ```
> ...
> 118       <family>Noto Sans Tagalog</family>
> 119       <family>Noto Sans Tagbanwa</family>
> 120       
> 121       <family>Noto Sans Tai Tham</family>
> 122       <family>Noto Sans Tai Viet</family>
> ...
> ```

Cool!!! Perfect solution it is!

Don't forget log-out and re-login.

**Comment by wipstones** [comment]
@devymex Found another font `Noto Sans Yi` which also covers some Chinese punctuation from `Noto Sans CJK`. However, the difference is only in the glyph design, so it won't cause the display issue. If you want all the displayed Chinese punctuation to be from `Noto Sans CJK`, it can be commented out, too.

**Comment by homanw104** [comment]
> @homanw104 I found that directly adding `/etc/fonts/local.conf` may cause some system UI fonts to change, such as the clock in top bar.

I acknowledged the clock font issue as well, but I think it's another issue on Gnome's side because they somehow use a non-standard character (RATIO U+2236) instead of a colon for the clock ([see the thread on Reddit](https://www.reddit.com/r/gnome/comments/t0a930/my_clock_thing_has_too_much_space_in_between_the/)). You'll need a font that has that RATIO character to set as the fallback font. Therefore the fix to this problem is dependent on your need.

Personally, I added *Noto Sans Math* right after *Noto Sans* to keep the fallback order as vanilla as it can be.

```xml
    <prefer>
      <family>Noto Sans</family>
      <family>Noto Sans Math</family>
      ...
    </prefer>
```

I recommend not to modify the configurations that come with the system as it may break on system upgrades. And commenting out existing fonts may also cause some characters not to render in some rare cases (like what if someone created a text emoji out of those?).

**Comment by Jobcrazy** [comment]
Why this issue is closed? It still exists in Ubuntu 24.0.2.

**Comment by hkurokawa** [comment]
Just for your information. I ran into this issue recently with Ubuntu 25.04 and I think I found a better solution.

You can create `~/.config/fontconfig/fonts.conf` with the following content (please replace `Noto Sans CJK JP` with whatever font you want to use for punctuations):
```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
<!--
  $XDG_CONFIG_HOME/fontconfig/fonts.conf for per-user font configuration
-->
<fontconfig>
  <match>
    <test name="lang" compare="contains">
      <string>en</string>
    </test>
    <test name="family">
      <string>sans-serif</string>
    </test>
    <edit name="family" mode="prepend">
      <string>Noto Sans</string>
    </edit>
    <edit name="family" mode="prepend">
      <string>Noto Sans Math</string>
    </edit>
    <edit name="family" mode="prepend">
      <string>Noto Sans CJK JP</string>
    </edit>
  </match>
</fontconfig>
```
Then, log out and log in again. The punctuation problem should be fixed and the clock also should look correct.

## More details

I think the issue here is that if a punctuation character (`U+3001`) is surrounded by ASCII-only characters (e.g., `abc、def`), the system assumes it is English and fontconfig looks for a font for English. You can run `fc-match -s :lang=en | head` to see the preference order of fonts. As is pointed out in [the above comment](https://github.com/canonical/Ubuntu-Sans-fonts/issues/117#issuecomment-2439368846), `Noto Sans Monglian` is at a higher place than `Noto Sans CJK JP` and it is probably used to render U+3001.

My solution is telling fontconfig to prefer `Noto Sans` → `Noto Sans Math` (for the clock) → `Noto Sans CJK JP` when the language is English.

Please feel free to let me know if you have any comments or questions.
